framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,605 @@
+     *     {@code TypeMirror}s that represent their base Java types
+     * @param types a map from fully-qualified names to the {@code ATypeElement}s representing their
+     *     declarations
+     * @param enumNamesToEnumConstant a map from fully-qualified enum names to the enum constants
+     *     defined in that name
+     * @param filename the path of the file to write to
+     * @throws IOException if the file doesn't exist
+     * @see #write(AScene, Map, Map, Map, Writer)
+     */
+    public static void write(
+            AScene scene,
+            Map<String, TypeMirror> basetypes,
+            Map<@FullyQualifiedName String, TypeElement> types,
+            Map<@FullyQualifiedName String, List<VariableElement>> enumNamesToEnumConstant,
+            String filename)
+            throws IOException {
+        write(scene, basetypes, types, enumNamesToEnumConstant, new FileWriter(filename));
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the package.
+     *
+     * @param className the binary name of a class
+     * @return the part of the name referring to the package
+     */
+    private static String packagePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        return (lastdot == -1) ? "" : className.substring(0, lastdot);
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the basename of the class. This method
+     * replaces the {@code $}s in the names of inner classes with {@code .}s, so that they can be
+     * printed correctly in stub files.
+     *
+     * @param className a binary name
+     * @return the part of the name representing the class's name without its package
+     */
+    private static String basenamePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        String result = (lastdot == -1) ? className : className.substring(lastdot + 1);
+        return result.replace('$', '.');
+    }
+
+    /**
+     * Formats a literal argument of an annotation. Similar to {@link IndexFileWriter#printValue},
+     * but does not print directly. Instead, returns the result to be printed.
+     *
+     * @param aft the annotation whose values are being formatted, for context
+     * @param o the value or values to format
+     * @return the String representation of the value
+     */
+    private String formatAnnotationValue(AnnotationFieldType aft, Object o) {
+        if (aft instanceof AnnotationAFT) {
+            return formatAnnotation((Annotation) o);
+        } else if (aft instanceof ArrayAFT) {
+            StringBuilder result = new StringBuilder();
+            ArrayAFT aaft = (ArrayAFT) aft;
+            result.append('{');
+            if (!(o instanceof List)) {
+                result.append(formatAnnotationValue(aaft.elementType, o));
+            } else {
+                List<?> l = (List<?>) o;
+                // watch out--could be an empty array of unknown type
```

Comment:
I don't understand this comment.  Who should watch out?  What should they do if they notice something?  Under what circumstances is it an empty array, and how is that related to null?

Reply:
I also don't understand it, but left it because this method is copied from the Annotation File Utilities, with modifications by me to make it functional rather than side-effect-ful. I can remove the comment, but I'd rather not touch this code unless we know it to be broken.

Reply:
If you copy a method, please write a comment to that effect.  It clarifies blame, but more importantly it helps us keep the two versions in sync.  (Can you not use the existing version?)

Reply:
I could not use the existing version, because it works by side-effecting a local field containing the writer, so it's useless in isolation. I've updated the comment on the method to be more clear, but it already mentioned that the code came from the Annotation File Utilities, so I'm not sure what else to do.